### PR TITLE
Matrix free module

### PR DESCRIPTION
**Pull request by kronbichler**
This new (pretty extensive already) module describes what I think are the most important components of the matrix-free framework in deal.II. I'm happy if someone wants to add something to this description. As we discussed with @Rombur and @bangerth in #6210, the module is currently mostly about the internal data structures. I hope this helps in reviewing #6300 and #6285 (and the follow-up) by a more abstract view.

### CODE REVIEW

**Comment by kronbichler** [comment]
Relates to #5667.

File: doc/doxygen/headers/matrixfree.h [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+ * individual field is of type
+ * internal::MatrixFreeFunctions::MappingInfoStorage and holds arrays with the
+ * inverse Jacobians, the JxW values, normal vectors, normal vectors times
+ * inverse Jacobians (for FEEvaluationAccess::get_normal_derivative()),
+ * quadrature points in real space, and quadrature points on the reference
+ * element. We use an auxiliary index array that points to the start of the
+ * data for each cell, namely the `data_index_offsets` field for the
+ * Jacobians, JxW values, and normal vectors, and `quadrature_point_offsets`
+ * for the quadrature points. This offset enables hp adaptivity with variable
+ * lenghts of fields similar to what is done for DoFInfo, but it also enables
```
Comment (Rombur) [inline doc/doxygen/headers/matrixfree.h]: typo lengths

File: doc/doxygen/headers/matrixfree.h [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+ * Jacobian, without the quadrature weight. As a consequence, we need to be
+ * careful in FEEvaluationBase::submit_value() and similar for this case as we
+ * must still multiply by the weight.
+ * <li> Affine cells have constant Jacobian within the whole cell, so only a
+ * single field needs to be stored per cell. Due to the similarity within the
+ * cell, we also check for other cell batches with the same Jacobian for all
+ * cells on the current processor. Since the JxW values in the general case
+ * store the Jacobian times the quadrature weight, but we only want to keep a
+ * single field for an affine cell, we misuse the name <i>JxW</i> in the
+ * Cartesian case and only store the determinant of the Jacobian, without the
```
Comment (Rombur) [inline doc/doxygen/headers/matrixfree.h]: Copy/paste mistake. We are in the affine case.

File: doc/doxygen/headers/matrixfree.h [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+ * The implementation of MappingInfo is split into cell and face parts, so the
+ * two components can be easily held apart. What makes the code a bit awkward
+ * to read is the fact that we need to batch several objects together from the
+ * original scalar evaluation done in an FEValues object, that we need to
+ * identify data fields that are repetitive, and that we need to fine the
+ * compression over several cells throught an `std::map` for the Cartesian and
+ * affine cases.
+ *
+ * The data computation part of MappingInfo is parallelized by tasks besides
+ * the obvious MPI parallelization. Each processor computers the information
```
Comment (Rombur) [inline doc/doxygen/headers/matrixfree.h]: typo computes

**Comment by kronbichler** [comment]
Thanks for the careful read. I've fixed those issues.

**Pull request merged**